### PR TITLE
Use Txid in COutpoint

### PR DESCRIPTION
**Pull request by dergoegge**
This PR changes the type of the hash of a transaction outpoint from `uint256` to `Txid`.

### CODE REVIEW

File: src/rest.cpp [hunk 1]
```diff
@@ -789,16 +789,14 @@ static bool rest_getutxos(const std::any& context, HTTPRequest* req, const std::
...
             int32_t nOutput;
             std::string strTxid = uriParts[i].substr(0, uriParts[i].find('-'));
             std::string strOutput = uriParts[i].substr(uriParts[i].find('-')+1);
 
             if (!ParseInt32(strOutput, &nOutput) || !IsHex(strTxid))
                 return RESTERR(req, HTTP_BAD_REQUEST, "Parse error");
 
-            txid.SetHex(strTxid);
-            vOutPoints.emplace_back(txid, (uint32_t)nOutput);
+            vOutPoints.emplace_back(TxidFromString(strTxid), (uint32_t)nOutput);
```
Comment (Sjors) [inline src/rest.cpp:799]: (Followup) should check that the hex string is of the correct length

File: src/rpc/mempool.cpp [hunk 2]
```diff
@@ -633,7 +633,7 @@ static RPCHelpMan gettxspendingprevout()
                                     {"vout", UniValueType(UniValue::VNUM)},
                                 }, /*fAllowNull=*/false, /*fStrict=*/true);
 
-                const uint256 txid(ParseHashO(o, "txid"));
+                const Txid txid = Txid::FromUint256(ParseHashO(o, "txid"));
```
Comment (Sjors) [inline src/rpc/mempool.cpp:636]: (Followup) should check length

File: src/primitives/transaction.h [hunk 3]
```diff
@@ -43,8 +43,7 @@ class COutPoint
 
     friend bool operator<(const COutPoint& a, const COutPoint& b)
     {
-        int cmp = a.hash.Compare(b.hash);
-        return cmp < 0 || (cmp == 0 && a.n < b.n);
+        return a.hash < b.hash || (a.hash == b.hash && a.n < b.n);
```
[the remark this answers is not in the document]
Comment (dergoegge) [reply src/primitives/transaction.h]: This is great, thanks!

File: src/rpc/blockchain.cpp [hunk 4]
```diff
@@ -2014,7 +2014,7 @@ bool FindScriptPubKey(std::atomic<int>& scan_progress, const std::atomic<bool>&
         }
         if (count % 256 == 0) {
             // update progress reference every 256 item
-            uint32_t high = 0x100 * *key.hash.begin() + *(key.hash.begin() + 1);
+            uint32_t high = 0x100 * *UCharCast(key.hash.begin()) + *(UCharCast(key.hash.begin()) + 1);
```
Comment (Sjors) [inline src/rpc/blockchain.cpp:2017]: This seems unrelated?

File: src/rpc/rawtransaction_util.cpp [hunk 5]
```diff
@@ -34,7 +34,7 @@ void AddInputs(CMutableTransaction& rawTx, const UniValue& inputs_in, std::optio
         const UniValue& input = inputs[idx];
         const UniValue& o = input.get_obj();
 
-        uint256 txid = ParseHashO(o, "txid");
+        Txid txid = Txid::FromUint256(ParseHashO(o, "txid"));
```
Comment (Sjors) [inline src/rpc/rawtransaction_util.cpp:37]: (Followup) should check length

File: src/rpc/rawtransaction_util.cpp [hunk 6]
```diff
@@ -185,7 +185,7 @@ void ParsePrevouts(const UniValue& prevTxsUnival, FillableSigningProvider* keyst
                     {"scriptPubKey", UniValueType(UniValue::VSTR)},
                 });
 
-            uint256 txid = ParseHashO(prevOut, "txid");
+            Txid txid = Txid::FromUint256(ParseHashO(prevOut, "txid"));
```
Comment (Sjors) [inline src/rpc/rawtransaction_util.cpp:188]: (Followup) should check length

File: src/wallet/rpc/coins.cpp [hunk 7]
```diff
@@ -320,7 +320,7 @@ RPCHelpMan lockunspent()
                 {"vout", UniValueType(UniValue::VNUM)},
             });
 
-        const uint256 txid(ParseHashO(o, "txid"));
+        const Txid txid = Txid::FromUint256(ParseHashO(o, "txid"));
```
Comment (Sjors) [inline src/wallet/rpc/coins.cpp:323]: (Followup) should check length

File: src/wallet/rpc/spend.cpp [hunk 8]
```diff
@@ -672,7 +672,7 @@ void FundTransaction(CWallet& wallet, CMutableTransaction& tx, CAmount& fee_out,
 
     if (options.exists("input_weights")) {
         for (const UniValue& input : options["input_weights"].get_array().getValues()) {
-            uint256 txid = ParseHashO(input, "txid");
+            Txid txid = Txid::FromUint256(ParseHashO(input, "txid"));
```
Comment (Sjors) [inline src/wallet/rpc/spend.cpp:675]: (Followup) should check length

File: src/wallet/spend.cpp [hunk 9]
```diff
@@ -321,7 +321,7 @@ CoinsResult AvailableCoins(const CWallet& wallet,
     std::set<uint256> trusted_parents;
     for (const auto& entry : wallet.mapWallet)
     {
-        const uint256& wtxid = entry.first;
+        const uint256& txid = entry.first;
```
Comment (Sjors) [inline src/wallet/spend.cpp:324]: (Followup) should check length

File: src/rpc/blockchain.cpp [hunk 10]
```diff
@@ -1057,7 +1057,7 @@ static RPCHelpMan gettxout()
 
     UniValue ret(UniValue::VOBJ);
 
-    uint256 hash(ParseHashV(request.params[0], "txid"));
+    auto hash{Txid::FromUint256(ParseHashV(request.params[0], "txid"))};
```
Comment (Sjors) [inline src/rpc/blockchain.cpp:1060]: `Txid hash` ?

**Review by Sjors — APPROVED** [review]
ACK 9e58c5bcd96e7ff2062274868814ccae0626589e

**Comment by Sjors** [comment]
@maflcko perhaps we should add another `RPCArg` type that's 32 characters `STR_HEX`?

File: src/rpc/blockchain.cpp (hunk 4 above)
Comment (dergoegge) [reply src/rpc/blockchain.cpp:2017]: `Txid::begin` etc. return `std::byte*` which can't be used for arithmetic operations.

See https://en.cppreference.com/w/cpp/types/byte.

File: src/rpc/mempool.cpp (hunk 2 above)
Comment (dergoegge) [reply src/rpc/mempool.cpp:636]: Looks like `ParseHashO` has a length check internally?

File: src/wallet/spend.cpp (hunk 9 above)
Comment (dergoegge) [reply src/wallet/spend.cpp:324]: Not sure what you mean here? `entry.first` is a `uint256`.

File: src/rpc/blockchain.cpp (hunk 4 above)
Comment (Sjors) [reply src/rpc/blockchain.cpp:2017]: I was confused where `Txid` comes to play here. But I see `key` is a `COutPoint`.

File: src/wallet/spend.cpp (hunk 9 above)
Comment (Sjors) [reply src/wallet/spend.cpp:324]: I think I copy-pasted this comment in the wrong file. Can be ignored

File: src/rpc/mempool.cpp (hunk 2 above)
Comment (Sjors) [reply src/rpc/mempool.cpp:636]: Ah, it calls `ParseHashV` which indeed checks that it's 64 characters. So then only the `rest.cpp` line above doesn't check the length (and it's terribly documented).

File: src/rpc/rawtransaction_util.cpp (hunk 5 above)
Comment (Sjors) [reply src/rpc/rawtransaction_util.cpp:37]: Update: it does, via `ParseHashV` - deleting the other comments below

**Comment by ismaelsadeeq** [comment]
Concept ACK

File: src/util/transaction_identifier.h
Comment (pablomartin4btc) [inline src/util/transaction_identifier.h:1]: Are you keeping this and removing it in a follow-up?

https://github.com/bitcoin/bitcoin/blob/9e58c5bcd96e7ff2062274868814ccae0626589e/src/util/transaction_identifier.h#L53-L60

Same for the comparisons above those lines (#L18 `// TODO`).

**Review by pablomartin4btc — COMMENTED** [review]
Concept ACK

Light CR.

File: src/test/fuzz/tx_pool.cpp [hunk 11]
```diff
@@ -367,15 +367,15 @@ FUZZ_TARGET(tx_pool, .init = initialize_tx_pool)
...
+    std::vector<Txid> txids;
     txids.reserve(g_outpoints_coinbase_init_mature.size());
     for (const auto& outpoint : g_outpoints_coinbase_init_mature) {
         txids.push_back(outpoint.hash);
     }
     for (int i{0}; i <= 3; ++i) {
         // Add some immature and non-existent outpoints
         txids.push_back(g_outpoints_coinbase_init_immature.at(i).hash);
-        txids.push_back(ConsumeUInt256(fuzzed_data_provider));
+        txids.push_back(Txid::FromUint256(ConsumeUInt256(fuzzed_data_provider)));
```
Comment (stickies-v) [inline src/test/fuzz/tx_pool.cpp:378]: nit
```suggestion
        txids.emplace_back(Txid::FromUint256(ConsumeUInt256(fuzzed_data_provider)));
```

File: src/test/miniminer_tests.cpp [hunk 12]
```diff
@@ -590,9 +590,17 @@ BOOST_FIXTURE_TEST_CASE(calculate_cluster, TestChain100Setup)
...
+        std::vector<uint256> out;
+        std::transform(vec.begin(), vec.end(), std::back_inserter(out),
+                       [](const Txid& txid) { return txid.ToUint256(); });
+        return out;
+    };
+
     // Add chain of size 500
     TestMemPoolEntryHelper entry;
-    std::vector<uint256> chain_txids;
+    std::vector<Txid> chain_txids;
```
Comment (stickies-v) [inline src/test/miniminer_tests.cpp:603]: Doesn't it make more sense to keep this as `std::vector<uint256> chain_txids;` for now?

File: src/wallet/spend.cpp (hunk 9 above)
Comment (stickies-v) [inline src/wallet/spend.cpp:324]: nit: Would simplify this to

```cpp
for (const auto& [txid, wtx] : wallet.mapWallet)
```

<details>
<summary>git diff on 9e58c5bcd9</summary>

```diff
diff --git a/src/wallet/spend.cpp b/src/wallet/spend.cpp
index 35583642a5..0e72efe0ba 100644
--- a/src/wallet/spend.cpp
+++ b/src/wallet/spend.cpp
@@ -319,11 +319,8 @@ CoinsResult AvailableCoins(const CWallet& wallet,
     std::vector<COutPoint> outpoints;
 
     std::set<uint256> trusted_parents;
-    for (const auto& entry : wallet.mapWallet)
+    for (const auto& [txid, wtx] : wallet.mapWallet)
     {
-        const uint256& txid = entry.first;
-        const CWalletTx& wtx = entry.second;
-
         if (wallet.IsTxImmatureCoinBase(wtx) && !params.include_immature_coinbase)
             continue;
 

```
</details>

**Review by stickies-v — APPROVED** [review]
ACK 9e58c5bcd96e7ff2062274868814ccae0626589e. A sizeable diff, but very straightforward changes. Didn't see anything controversial. Left a few nits, but nothing blocking, only if you have to retouch.

File: src/test/fuzz/mini_miner.cpp [hunk 13]
```diff
@@ -25,7 +25,7 @@ void initialize_miner()
     static const auto testing_setup = MakeNoLogFileContext<const TestingSetup>();
     g_setup = testing_setup.get();
     for (uint32_t i = 0; i < uint32_t{100}; ++i) {
-        g_available_coins.emplace_back(uint256::ZERO, i);
+        g_available_coins.emplace_back(Txid::FromUint256(uint256::ZERO), i);
```
Comment (TheCharlatan) [inline src/test/fuzz/mini_miner.cpp:28]: Nit: Just do `Txid{}`, like you did in `bench/disconnected_transactions`?

**Review by sedited — APPROVED** [review]
ACK 9e58c5bcd96e7ff2062274868814ccae0626589e

File: src/test/fuzz/tx_pool.cpp (hunk 11 above)
Comment (stickies-v) [reply src/test/fuzz/tx_pool.cpp:378]: nvm, I don't think this is actually an improvement.

File: src/util/transaction_identifier.h
Comment (dergoegge) [reply src/util/transaction_identifier.h:1]: Yes the plan is to remove it eventually, but it should still be needed at the moment.

File: src/rpc/txoutproof.cpp [hunk 14]
```diff
@@ -69,7 +69,7 @@ static RPCHelpMan gettxoutproof()
 
                 // Loop through txids and try to find which block they're in. Exit loop once a block is found.
                 for (const auto& tx : setTxids) {
-                    const Coin& coin = AccessByTxid(active_chainstate.CoinsTip(), tx);
+                    const Coin& coin = AccessByTxid(active_chainstate.CoinsTip(), Txid::FromUint256(tx));
```
Comment (achow101) [inline src/rpc/txoutproof.cpp:72]: New warning (fails with `-werror`)

```
rpc/txoutproof.cpp: In lambda function:
rpc/txoutproof.cpp:72:33: error: possibly dangling reference to a temporary [-Werror=dangling-reference]
   72 |                     const Coin& coin = AccessByTxid(active_chainstate.CoinsTip(), Txid::FromUint256(tx));
      |                                 ^~~~
rpc/txoutproof.cpp:72:52: note: the temporary was destroyed at the end of the full expression ‘AccessByTxid((*(const CCoinsViewCache*)(&(& active_chainstate)->Chainstate::CoinsTip())), transaction_identifier<false>::FromUint256((* & tx)))’
   72 |                     const Coin& coin = AccessByTxid(active_chainstate.CoinsTip(), Txid::FromUint256(tx));
      |                                        ~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
cc1plus: all warnings being treated as errors
```

Comment (dergoegge) [reply src/rpc/txoutproof.cpp:72]: What gcc version are you using? I don't get this warning (gcc 12).

Comment (achow101) [reply src/rpc/txoutproof.cpp:72]: gcc (GCC) 13.2.1 20230801
